src/QsCompiler/Transformations/Monomorphization.cs

```diff
@@ -6,47 +6,90 @@
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Linq;
+using System.Runtime;
+using System.Runtime.InteropServices;
 using Microsoft.Quantum.QsCompiler.DataTypes;
+using Microsoft.Quantum.QsCompiler.Optimizations;
+using Microsoft.Quantum.QsCompiler.SymbolManagement;
 using Microsoft.Quantum.QsCompiler.SyntaxTokens;
 using Microsoft.Quantum.QsCompiler.SyntaxTree;
 using Microsoft.Quantum.QsCompiler.Transformations.Core;
 
 namespace Microsoft.Quantum.QsCompiler.Transformations.Monomorphization
 {
+    using LocalTypeKind = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>;
+    using GenericsCrate = Dictionary<QsQualifiedName, (QsCallable, Dictionary<Concretion, QsCallable>)>;
+
+    #region Monomorphization
+
+    internal class Concretion : HashSet<(NonNullable<string>, LocalTypeKind)>
+    {
+        public override int GetHashCode()
+        {
+            return this.Select(x => (x.Item1.Value.GetHashCode(), NamespaceManager.TypeHash(ResolvedType.New(x.Item2))).ToString().GetHashCode())
+                .Aggregate((sum, i) => unchecked(sum + i)); // TODO: Use better hashing over enumerable
+        }
+
+        public override bool Equals(object obj)
+        {
+            return this.GetHashCode() == obj.GetHashCode();
+        }
+    }
+
     public class ResolveGenericsSyntax :
         SyntaxTreeTransformation<ScopeTransformation<ResolveGenericsExpression>>
     {
+        private GenericsCrate Generics;
+
         public static IEnumerable<QsNamespace> Apply(IEnumerable<QsNamespace> namespaces)
         {
             if (namespaces == null || namespaces.Contains(null)) throw new ArgumentNullException(nameof(namespaces));
 
             // Get list of generics
-            List<QsQualifiedName> generics = new List<QsQualifiedName>();
-            var Callables = namespaces.GlobalCallableResolutions();
-            foreach (var callable in Callables)
-            {
-                if (callable.Value.Signature.TypeParameters.Any())
-                {
-                    generics.Add(callable.Key);
-                }
-            }
+            // TODO: this is specific to Provided implementations, but should be general to all implementation kinds
+            GenericsCrate generics = namespaces.GlobalCallableResolutions()
+                .Where(call => call.Value.Signature.TypeParameters.Any() && call.Value.Specializations.Any(x => x.Implementation.IsProvided))
+                .ToDictionary(call => call.Key, call => (call.Value, new Dictionary<Concretion, QsCallable>()));
 
-            var filter = new ResolveGenericsSyntax(new ScopeTransformation<ResolveGenericsExpression>(new ResolveGenericsExpression()));
+            var filter = new ResolveGenericsSyntax(new ScopeTransformation<ResolveGenericsExpression>(new ResolveGenericsExpression(generics)), generics);
```

Comment:
Instead of handing down the GenericsCrate instance you might consider handing it a function that given the full name of a callable returns the dictionary of concretions.

Reply:
The GenericsCreate class and its instances will no longer be used.